{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "The below code document are tips and tricks for common downstream\n",
    "analysis next steps. Each section contains information on how to perform\n",
    "scRNA-seq analysis, including identification of low quality cells,\n",
    "merging and removal of clusters/cells, reclustering, changing cell type\n",
    "annotations and identities, etc.\n",
    "\n",
    "Please contact support@honeycomb.bio for all questions.\n",
    "\n",
    "# Package Installation and library loading\n",
    "\n",
    "The below packages needs to be installed prior to running this vignette.\n",
    "Code for package installation is below. There may be additional\n",
    "dependencies required by these packages you may have to install\n",
    "depending on your R environment.\n",
    "`{r packageinstall, eval=F} install.packages(\"htmltools\") install.packages(\"tidyverse\") install.packages(\"patchwork\") install.packages(\"plyr\") install.packages('Seurat') if (!require(\"BiocManager\", quietly = TRUE)){   install.packages(\"BiocManager\") } BiocManager::install(\"UCell\") install.packages(\"ggpubr\") BiocManager::install(\"SingleCellExperiment\") BiocManager::install(\"scuttle\") BiocManager::install(\"celldex\") install.packages(\"DT\") if (!requireNamespace(\"remotes\", quietly = TRUE)) {   install.packages(\"remotes\") } install.packages(\"HGNChelper\")`\n",
    "Load libraries using the code below:\n",
    "`{r libraries, warning = FALSE} suppressPackageStartupMessages({   library(tidyverse)   library(patchwork)   library(plyr)   library(Seurat)   library(UCell)   library(ggpubr)   library(SingleCellExperiment)   library('scuttle')   library(celldex)   library(DT)   library(scales)   library(HGNChelper) })`\n",
    "\n",
    "# Copying BeeNet/BeeNetPLUS output files from Google Cloud\n",
    "\n",
    "The code below loads your R object from the Google Cloud Platform\n",
    "Bucket. You will need to replace code where outlined below: \\`\\`\\`{r\n",
    "copyfiles, eval = F, echo = T}\n",
    "\n",
    "system(“gsutil -m cp -r\n",
    "gs://resources.honeycomb.bio/honeycomb-public-scripts/HC-logo.png .”)\n",
    "system(“mkdir -p data”)\n",
    "\n",
    "#Replace the URL with the link to the R object output by BeeNetPLUS or\n",
    "your own downstream R object. Navigate to your Google Cloud output\n",
    "directory and locate the .tgz object or R object and copy the path into\n",
    "the code below:\n",
    "\n",
    "system(“gsutil -m cp\n",
    "gs://hcb-external-pipeline-testing/20230426-TestData/Analysis-20230426-300-600/Analysis-20230426-300-600.tgz\n",
    "./data”)\n",
    "\n",
    "#Unzip the file if it is a .tgz file system(“tar -zxf data/\\*tgz -C\n",
    "./data”)\n",
    "\n",
    "    Next, load your R object into the environment. This tutorial is based on an R object named \"obj\". For **BeeNetPLUSv1.0** users, change the name of the object using the code below. \n",
    "    ```{r loadobject}\n",
    "    #Replace the file path for the R object and load in the R object\n",
    "    load(\"./data/Analysis-20230426-300-600/FullDataset-20230426-TestData.Rdata\")\n",
    "\n",
    "    #This should load a Seurat object called umi_seurat and several other data frames and variables that contain pipeline parameters and other quality control metrics. \n",
    "\n",
    "    # Rename the umi_seurat object obj for this analysis \n",
    "    obj<- umi_seurat\n",
    "    rm(umi_seurat)\n",
    "\n",
    "`{r setup, include=FALSE} knitr::opts_chunk$set(echo = TRUE) knitr::opts_chunk$set(root.dir = \"./data\")`\n",
    "\n",
    "# Metadata Tips and Tricks\n",
    "\n",
    "## Collecting metadata from your Seurat Object\n",
    "\n",
    "\\`\\`\\` {r metadata} \\# The code below creates a data frame of metadata\n",
    "included in the Seurat object metadata \\<-\n",
    "rownames_to_column(obj@meta.data, “cellids”)\n",
    "\n",
    "datatable(metadata)\n",
    "\n",
    "\n",
    "    ## Setting the Ident of Cells\n",
    "    You can set the Cell Identity to any column in the object metadata. After clustering, the Ident is automatically the cluster number, but this can be changed to cell type, sample name, or group depending on the downstream analysis to be performed.\n",
    "    ``` {r clustering, warning=FALSE, message=FALSE}\n",
    "    ##Set the Identity of the cells to the seurat cluster\n",
    "    Idents(obj)<- obj$seurat_clusters\n",
    "    DimPlot(obj, reduction=\"umap\")\n",
    "\n",
    "## Changing the Order of Idents\n",
    "\n",
    "Change the “level” of the metadata parameter “seurat_clusters” so that\n",
    "the numbers are in order. Releveling can be performed for any metadata\n",
    "and changed at any time. \\`\\`\\` {r releveling, warning=FALSE, message\n",
    "=FALSE} obj$seurat_clusters<- factor(obj$seurat_clusters, levels =\n",
    "c(“1”, “2”, “3”, “4”, “5”, “6”, “7”, “8”, “9”, “10”, “11”, “12”, “13”,\n",
    "“14”, “15”, “16”, “17”, “18”)) Idents(obj)\\<- obj\\$seurat_clusters\n",
    "\n",
    "DimPlot(obj, reduction=“umap”)\n",
    "\n",
    "\n",
    "    # Clustering Tips and Tricks\n",
    "    ## Identifying and Removing Low Quality Clusters\n",
    "    A high quality UMAP contains:  \n",
    "\n",
    "    - Clusters that are distinct from one another.  \n",
    "    - Clusters that are not sample specific (unless biologically relevant).  \n",
    "\n",
    "    ``` {r HQclusterhighlight, warning=FALSE, message =FALSE}\n",
    "    DimPlot(obj,pt.size = 0.2, cells.highlight=WhichCells(obj, idents = 4), cols.highlight = \"orange\", cols= \"grey\")\n",
    "\n",
    "A low quality cluster on a UMAP may:\n",
    "\n",
    "-   Be diffuse around the UMAP. These clusters may be defined by RBC\n",
    "    contamination, high mitochondrial gene expression, or abnormally low\n",
    "    genes/transcripts.  \n",
    "-   Be in the middle of or bridge two larger, unrelated clusters. This\n",
    "    could be indicative of multiplets.\n",
    "\n",
    "`{r LQclusterhighlight, warning=FALSE, message =FALSE} DimPlot(obj,pt.size = 0.2, cells.highlight=WhichCells(obj, idents = 12), cols.highlight = \"orange\", cols= \"grey\")`\n",
    "\n",
    "In this example, we think Cluster 12 might be low quality due to its\n",
    "positioning on the UMAP. We also want to check to see if there are\n",
    "distinct marker genes that define cluster 12, as well as the overall\n",
    "numbers of genes, transcripts, and mitochondrial reads\n",
    "\n",
    "\\`\\`\\` {r LQQC, warning=FALSE, message =FALSE}\n",
    "\n",
    "#Get the median QC parameters for the whole object\n",
    "median(obj$nCount_RNA) median(obj$nFeature_RNA)\n",
    "median(obj\\$percent.mito)\n",
    "\n",
    "#Isolate c12 and get the QC parameters c12\\<- subset(obj, subset =\n",
    "seurat_clusters==“12”) median(c12$nCount_RNA) median(c12$nFeature_RNA)\n",
    "median(c12\\$percent.mito)\n",
    "\n",
    "#Find Markers for just c12 markers12 \\<- FindMarkers(object = obj,\n",
    "ident.1 = 12) datatable(markers12\\[1:5,\\])\n",
    "\n",
    "\n",
    "    As cluster 12 does not isolate on its own, has no biologically significant marker genes (log2fc <-0.58 or >0.58) and has significantly lower genes and transcripts per cell than the rest of the object, this cluster is a good candidate for removal. Please note that certain cell types (e.g. granulocytes) have lower numbers of genes and transcripts than others, and therefore looking at marker genes and UMAP positioning is necessary to help determine whether a cluster should be removed.\n",
    "\n",
    "    ``` {r c12remove, warning=FALSE, message =FALSE}\n",
    "\n",
    "    #Remove Cluster 12\n",
    "    obj<- subset(obj, subset = seurat_clusters !=\"12\")\n",
    "\n",
    "    #Recluster\n",
    "    obj <- NormalizeData(obj)\n",
    "    obj <- ScaleData(obj)\n",
    "    obj <- FindVariableFeatures(obj, selection.method = \"vst\", nfeatures = 2000)\n",
    "    obj<-RunPCA(obj, verbose = FALSE)\n",
    "    obj<-RunUMAP(obj, dims=1:30,verbose = FALSE)\n",
    "    obj<- FindNeighbors(obj,verbose = FALSE)\n",
    "    obj<- FindClusters(obj, verbose = FALSE)\n",
    "    obj=BuildClusterTree(obj,reorder=TRUE,reorder.numeric=TRUE)\n",
    "\n",
    "    #Add scTransform as another option\n",
    "    obj <- SCTransform(obj, vars.to.regress = \"percent.mito\", verbose = FALSE)\n",
    "    obj<-RunPCA(obj, verbose = FALSE)\n",
    "    obj<-RunUMAP(obj, dims=1:18,verbose = FALSE)\n",
    "    obj<- FindNeighbors(obj,verbose = FALSE)\n",
    "    obj<- FindClusters(obj, verbose = FALSE)\n",
    "    obj=BuildClusterTree(obj,reorder=TRUE,reorder.numeric=TRUE)\n",
    "    obj$seurat_clusters<- obj$tree.ident\n",
    "\n",
    "    DimPlot(obj, reduction=\"umap\")\n",
    "\n",
    "Iterate through removing low quality clusters one at a time until high\n",
    "quality cells remain. Be sure to re-do any Violin Plots, QC\n",
    "plots/tables, and UMAPs with the updated Seurat object.\n",
    "\n",
    "\\`\\`\\` {r qcbysample, warning=FALSE, message=FALSE}\n",
    "\n",
    "#Redo the QC qc_metrics_sample\\<- function(obj){\n",
    "I=obj@meta.data$orig.ident  Reads=obj@meta.data$ExonReads\n",
    "ReadsM=obj@meta.data$reads.mapped  ReadsT=obj@meta.data$reads.Total\n",
    "trans=obj@meta.data$nCount_RNA  Genes=obj@meta.data$nFeature_RNA\n",
    "Mito=obj@meta.data$percent.mito  Complexity=obj@meta.data$Complexity\n",
    "nSamples=length(unique(obj$groupSamp))  readEC=vector(mode=\"double\",length=nSamples)  readsTot=totReads.df  readsTotalSamples=colnames(readsTot)  names(readEC)=sort(unique(obj$groupSamp))\n",
    "readTC=readEC readMC=readEC transC=readEC geneC=readEC mitoC=readEC\n",
    "nCells=readEC compl=readEC readsCellTotal=readEC readsSampleTotal=readEC\n",
    "FracCellReads=readEC readsSampleMapped=readEC readsSampleFiltered=readEC\n",
    "readsSampleExon=readEC percExon=readEC percFilt=readEC nF=readEC\n",
    "r=readEC v1=readEC v2=readEC v3=readEC date=readEC sT=readEC DS=readEC\n",
    "G=readEC T=readEC BC=readEC seqsat=readEC sampleName=readEC\n",
    "sampleGroup=readEC sampleslist\\<-sort(unique(obj\\$orig.ident))\n",
    "\n",
    "for (x in 1:nSamples) {\n",
    "readTC\\[x\\]=median(ReadsT\\[is.na(match(I,sampleslist\\[x\\]))==0\\])\n",
    "readEC\\[x\\]=median(Reads\\[is.na(match(I,sampleslist\\[x\\]))==0\\])\n",
    "readMC\\[x\\]=median(ReadsM\\[is.na(match(I,sampleslist\\[x\\]))==0\\])\n",
    "transC\\[x\\]=median(trans\\[is.na(match(I,sampleslist\\[x\\]))==0\\])\n",
    "geneC\\[x\\]=median(Genes\\[is.na(match(I,sampleslist\\[x\\]))==0\\])\n",
    "mitoC\\[x\\]=median(Mito\\[is.na(match(I,sampleslist\\[x\\]))==0\\])\n",
    "compl\\[x\\]=median(Complexity\\[is.na(match(I,sampleslist\\[x\\]))==0\\])\n",
    "nCells\\[x\\]=sum(is.na(match(I,sampleslist\\[x\\]))==0)\n",
    "readsCellTotal\\[x\\]=sum(ReadsT\\[is.na(match(I,sampleslist\\[x\\]))==0\\])\n",
    "sampleName\\[x\\]=sampleslist\\[x\\]\n",
    "readsSampleTotal\\[x\\]=readsTot\\[1,is.na(match(readsTotalSamples,sampleslist\\[x\\]))==0\\]\n",
    "FracCellReads\\[x\\]=readsCellTotal\\[x\\]/readsSampleTotal\\[x\\]\n",
    "readsSampleMapped\\[x\\]=readsTot\\[3,is.na(match(readsTotalSamples,sampleslist\\[x\\]))==0\\]\n",
    "readsSampleFiltered\\[x\\]=readsTot\\[2,is.na(match(readsTotalSamples,sampleslist\\[x\\]))==0\\]\n",
    "readsSampleExon\\[x\\]=readsTot\\[4,is.na(match(readsTotalSamples,sampleslist\\[x\\]))==0\\]\n",
    "percExon\\[x\\]=readEC\\[x\\]/readTC\\[x\\]\\*100\n",
    "percFilt\\[x\\]=readsSampleFiltered\\[x\\]/readsSampleTotal\\[x\\]\n",
    "date\\[x\\]=gsub(“-”,““,Sys.Date()) G\\[x\\]=nGeneI T\\[x\\]=nTranI\n",
    "seqsat\\[x\\]=1-(sum(trans\\[is.na(match(I,sampleslist\\[x\\]))==0\\] /\n",
    "sum(ReadsM\\[is.na(match(I,sampleslist\\[x\\]))==0\\]))) }\n",
    "\n",
    "clustInfo=data.frame(date,sampleName,G, T, readTC,\n",
    "readMC,readEC,transC,geneC,mitoC,nCells,\n",
    "compl,readsSampleTotal,readsSampleFiltered,\n",
    "readsSampleMapped,readsSampleExon,readsCellTotal,\n",
    "FracCellReads,percExon,percFilt, seqsat)\n",
    "\n",
    "colnames(clustInfo)=c(“AnalysisDate”, “SampleID”, “GeneThreshold”,\n",
    "“TranscriptThreshold”, “TotalReads”, “MappedReads”, “ExonReads”,\n",
    "“nTrans”, “nGenes”, “percMito” ,“nCells”, “Complexity”,\n",
    "“SampleTotalReads”, “SampleFilteredReads”, “SampleMappedReads”,\n",
    "“SampleExonReads”, “HQCellsTotalReads”, “FracReadsHQCells”, “PctExon”,\n",
    "“FracPassFilter”, “SeqSat”)\n",
    "\n",
    "write.table(clustInfo,“SamplesMetrics.txt”,sep=“,col.names=NA)\n",
    "\n",
    "return(clustInfo) }\n",
    "\n",
    "sample_metrics\\<- qc_metrics_sample(obj)\n",
    "\n",
    "datatable(sample_metrics)\n",
    "\n",
    "    ## Merging Clusters\n",
    "    If two or more clusters have the same marker genes that define them, with no distinct marker genes distinguishing between either of the two clusters (see below section for details on how to compare two clusters), these clusters can be merged to form one cluster. We do not recommend overwriting the original clustering; rather, we suggest making a new column in the metadata and reassigning clusters there.\n",
    "\n",
    "    ``` {r mergeclusters, warning=FALSE, message=FALSE}\n",
    "    #Copy the original clusters to a new column called new.cluster.ids, releveling to make sure they are in the correct order\n",
    "    obj$new.cluster.ids<- factor(obj$seurat_clusters, levels = c(\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\"))\n",
    "\n",
    "    #Set the cell identities to the new cell clusters \n",
    "    Idents(obj)<- obj$new.cluster.ids\n",
    "\n",
    "    #Set new Idents to combine original clusters 6 and 7 into one cluster (6), and then renumbering the rest so that they go in order\n",
    "    new.cluster.ids<- c(\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\")\n",
    "    names(new.cluster.ids) <- levels(obj)\n",
    "\n",
    "    obj <- RenameIdents(obj, new.cluster.ids)\n",
    "    DimPlot(obj)\n",
    "\n",
    "Note that when you rename, merge, or remove clusters, subsequent\n",
    "analyses will need to be repeated. Rerun plot and figure generation,\n",
    "marker gene analysis and cluster QC parameters with the new cluster\n",
    "numbers for consistency in your data!\n",
    "\n",
    "# Marker Analysis Tips and Tricks\n",
    "\n",
    "We recommend using a min.pct of 0.25 to find marker genes. This ensures\n",
    "that markers are frequently expressed in at least one group.\n",
    "\n",
    "## Find markers for just one cluster\n",
    "\n",
    "`{r oneclustermarker, warning=FALSE, message=FALSE} #Find markers for cluster 1 (ident.1) Idents(obj)<- obj$seurat_clusters cluster1.markers <- FindMarkers(object = obj, ident.1 = 1, min.pct = 0.25) datatable(cluster1.markers)`\n",
    "\\## Directly compare two clusters\n",
    "`{r onevs2clustermarker, warning=FALSE, message=FALSE} #Compare clusters 1 and 2 directly Idents(obj)<- obj$seurat_clusters cluster1v2.markers <- FindMarkers(object = obj, ident.1 = 1, ident.2 = 2, min.pct = 0.25) datatable(cluster1v2.markers)`\n",
    "\\## Compare cell types\n",
    "`{r bvmclustermarker, warning=FALSE, message=FALSE} #Compare markers in B cells versus Monocytes Idents(obj)<- obj$cell_type clusterbvm.markers <- FindMarkers(object = obj, ident.1 = \"Naive B cells\", ident.2 = \"Monocyte\", min.pct = 0.25) datatable(clusterbvm.markers)`\n",
    "\\## Compare cell types by group\n",
    "`{r celltypesbygroup, warning=FALSE, message=FALSE} #Compare markers in B cells in 7.5k group versus the 30k group Idents(obj)<- obj$cell_type group.markers <- FindMarkers(object = obj, ident.1 = \"7.5k\", ident.2 = \"30k\", group.by = \"group\", subset.ident = \"Naive B cells\", min.pct = 0.25) datatable(group.markers)`\n",
    "\n",
    "# Cell Type Annotation Tips and Tricks\n",
    "\n",
    "## Reassigning Cell Types\n",
    "\n",
    "We do not recommend overwriting the original cell type annotation;\n",
    "rather, we suggest making a new column in the metadata and reassigning\n",
    "cell types there. In this object, cells that were annotated as\n",
    "“Neutrophil-2” and “Neutrophil-3” are actually monocytes, based on their\n",
    "positioning on the UMAP and marker gene expression. Here we will clean\n",
    "up the data and reassign these cells. Be sure to re-perform all\n",
    "downstream analyses pertaining to cell types after reassignment!\n",
    "\n",
    "`{r reassign, warning=FALSE, message=FALSE} obj$cell_type_new<- obj$cell_type obj$cell_type_new[obj$cell_type == \"Neutrophil-2\"]<- \"Monocyte\" obj$cell_type_new[obj$cell_type == \"Neutrophil-3\"]<- \"Monocyte\" DimPlot(obj, group.by = \"cell_type_new\")`\n",
    "\n",
    "# Saving your final R object\n",
    "\n",
    "Be sure to save your final object! Include all metadata tables and\n",
    "environment variables that you want to include in your final saved R\n",
    "object Here we saved our object and metadata as a new file called\n",
    "FilteredDataset-<DATASETNAME>.Rdata.\n",
    "\n",
    "\\`\\`\\` {r save, warning=FALSE, message=FALSE} \\# The code below\n",
    "preserves environment variables for BeeNetPLUSv1.0.\n",
    "\n",
    "save(obj, totReads.df, dataNames, input, sampletype,\n",
    "known_markers,nFastq, len, refs, beeV, wdlV, dataset, nGeneI, nTranI,\n",
    "clustGroupCountsN, clustCountsN2, clustCountsN3, clustGroupCountsN2,\n",
    "file=paste0(“FilteredDataset-”, dataset, “.Rdata”))\n",
    "\n",
    "\n",
    "    Not all the variables above are present for BeeNet users, but consider at least saving the final Seurat object, for downstream use. The following code saves only the Seurat object. Any other variables, functions, or data frames made in this tutorial will have to be regenerated in future analysis if not saved. \n",
    "\n",
    "    ```{r save_BeeNetOnly, eval=FALSE}\n",
    "    save(obj, \"FilteredDataset-NAMEofDATASET-SeuratObjectOnly.Rdata)\n",
    "\n",
    "# R Session\n",
    "\n",
    "\\`\\`\\`{r SessionInfo, echo=FALSE, results = ‘asis’} df_session_platform\n",
    "\\<- devtools::session_info()\\$platform %\\>% unlist(.) %\\>%\n",
    "as.data.frame(.) %\\>% rownames_to_column(.)\n",
    "\n",
    "colnames(df_session_platform) \\<- c(“Setting”, “Value”)\n",
    "\n",
    "knitr::kable( df_session_platform, booktabs = T, align = “l”, caption =\n",
    "“R Session Info”, \\# complete caption for main document )\n",
    "\n",
    "\\`\\`\\` \\# Package Info\n",
    "\n",
    "\\`\\`\\`{r PackageInfo, echo=FALSE, results = ‘asis’} df_session_packages\n",
    "\\<- devtools::session_info()\\$packages %\\>% as.data.frame(.) %\\>%\n",
    "filter(attached == TRUE) %\\>% dplyr::select(loadedversion, date) %\\>%\n",
    "rownames_to_column\n",
    "\n",
    "colnames(df_session_packages) \\<- c(“Package”, “Loaded version”, “Date”)\n",
    "\n",
    "knitr::kable( df_session_packages, booktabs = T, align = “l”, caption =\n",
    "“R Package Info”, \\# complete caption for main document )"
   ],
   "id": "60e453ab-049a-436a-a208-464ba245ebea"
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
